### ISSUE TITLE
Que doesn't reconnect after PG::UnableToSend

### ISSUE THREAD

**Issue by chendo** [chanks/que#93]
que 0.9.1.

We've seen some cases where Que loses the connection to Postgres and we see log messages saying:

```
{"lib":"que","hostname":"<snip>","pid":37,"thread":69940584448760,"event":"job_errored","error":{"class":"PG::UnableToSend","message":""},"job":null}
```

A restart of the app is required to resolve this issue. Also, the `Que.error_handler` doesn't get called for this error, so the only way we noticed it was alerts from jobs not being run. We haven't tried 0.9.2 just yet as I only noticed the update when I came to file an issue, but 0.9.2 doesn't look like it has a fix for it.

**Comment by chanks** [comment]
Hmm, I don't see any reason the error handler wouldn't be called for this,
if you've defined one. Is it possible that it's raising an error itself? If
it is it'll be silently discarded, which is unfortunate but I don't really
know of a better way to handle it.

Also, what connection pool are you using? Que doesn't establish connections
itself, so reestablishing failed connections is kind of beyond its scope
and should be handled by whatever you're using to connect to the DB. This
would require Que to make the connection pool aware of the connection
error, and I'm not really sure of the best way to handle that. I suppose
we'd have to special-case the connection errors and reraise them after the
error handler runs, or possibly have a separate error handling workflow for
errors that happen as the job is retrieved.

**Comment by chendo** [comment]
Hmm, yeah I can't see how it wouldn't be either. Will try wrapping the error handler and checking.

We're using the `pg` gem in Rails 4.2.0. Rails itself seems to handle reconnections fine, but Que doesn't appear to recover for some reason. We haven't explicitly set the adapter yet (should we?).

For the record, we're using pgpool load balanced behind haproxy.

**Comment by chanks** [comment]
Explicitly setting the adapter wouldn't help. Does it appear that Que keeps
raising PG::UnableToSend even after ActiveRecord is able to start executing
queries again (assuming you're running them in the same process and they're
sharing a connection pool)? If so, I don't have an immediate explanation
for that.

Not familiar with pgpool or haproxy, but if they work like pg_bouncer,
you'll probably get issues of some kind (though not PG::UnableToSend, I'd
think) - Que relies on being able to use the same connection as it takes a
lock, completes a job, and then releases the lock (and not all in the
context of a transaction, either).

**Comment by chendo** [comment]
Yeah, it does. We have a similar issue where the connection times out due to inactivity when doing a `rails console`, and it never recovers. I'm hoping a fix for that will also fix que. I'll keep you posted

**Comment by chendo** [comment]
We tweaked pgpool config but now we occasionally see `ERROR: unable to bind DETAIL: cannot get parse message "que_check_job" ERROR: unable to bind DETAIL: cannot get parse message "que_check_job"` after the connection reconnects successfully.

It seems to be due to the new connection not having the prepared statement set up, and from what I can see, que will retry so it's all good, yeah?

**Comment by chendo** [comment]
Also it turns out the error handler was being called, but we always expected `job` to not be nil. Thanks!

**Comment by chendo** [comment]
Actually, it doesn't seem to recover correctly after the `que_check_job` error, and will then start erroring with:

```
{"lib":"que","hostname":"<snip>","pid":16227,"thread":70322654856720,"event":"job_errored","error":{"class":"PG::InternalError","message":"ERROR:  unable to bind\nDETAIL:  cannot get parse message \"que_lock_job\"\nERROR:  unable to bind\nDETAIL:  cannot get parse message \"que_lock_job\"\n"},"job":null}
```

**Comment by statianzo** [comment]
We had the issue of a Que worker hanging after a postgres maintenance restart last week. Que 0.9.0.

**Comment by chanks** [comment]
Still looking into how to fix this, but wanted to say that I looked into pgpool a bit, and like pg_bouncer, I doubt that Que will be able to work cleanly with it. Que relies on being able to check out a connection and use it to take a lock and then release that lock later, and with pgpool or pg_bouncer in between there's no guarantee that the connection you use to try to clear the lock will be the same as the one you used to take it. So locks will be left hanging, which I expect will take up an ever-increasing amount of memory in Postgres and also cause old failed jobs to never be retried. I may be wrong, I haven't actually tried it.

Not sure if that would be related to the UnableToSend error you're seeing.

**Comment by chendo** [comment]
`UnableToSend` might be related to something else, but the pgpool issue is causing connections to be marked as "still in use" and thus we run out of connections on pgpool at some point. I'll be looking for a fix for this today.

**Comment by chanks** [comment]
Oh, so pgpool recognizes that the lock is taken and doesn't let anything else use that connection? That's reassuring.

**Comment by chendo** [comment]
Yeah, same with transactions, however it appears ActiveRecord does not properly clear the locks when it checks them back into the pool or something, so this also might be a problem with ActiveRecord too

**Comment by chendo** [comment]
We seem to have fixed the issue by having our own adapter that puts the connection inside a transaction.

```
module Que
  module Adapters
    class PGPoolActiveRecord < ActiveRecord
      def checkout
        checkout_activerecord_adapter do |conn|
          conn.transaction do
            yield conn.raw_connection
          end
        end
      end
    end
  end
end
```

**Comment by chanks** [comment]
Just wanted to drop a quick note to say that I haven't forgotten this. It doesn't sound urgent anymore, but I want to keep it on my TODO list to ensure that Que doesn't interfere with Sequel/ActiveRecord's ability to recognize and recover from connection failures in general. It looks like it'll require digging into both of those libraries a bit, though, and I'm not sure offhand how testable it'll be. But I'll do what I can.

**Comment by swrobel** [comment]
@chanks I know this is a pretty old issue, but I'm using the latest master of que on my app and have seen this popping up in Bugsnag over the past few days. I'm not sure what, if anything, triggered it, but it doesn't seem to be affecting any behavior, so I'm not entirely sure if I should be concerned or not ... anyway, just wanted to check in and see if you ever made progress w/ it.

**Comment by chanks** [comment]
Nope, sorry, never found out what the issue was here. Also note that this issue was opened for 0.x, and 1.0 was basically a complete rewrite.

Also I've handed off support of the gem, but if you can pinpoint something specific that you changed around the time the issue started appearing (a postgres upgrade, a Rails/ActiveRecord upgrade...) I'm sure that'd help out anyone who winds up looking into this.
